File: store/ef/ef.go
// Package ef codes one posting chunk: a sorted list of txnums as a plain
// Elias-Fano sequence, followed by a fixed-width payload per entry.
//
// Layout: uvarint n | uvarint lowBits | uvarint payloadBits | upper bits
// (unary, n ones, ceil((max>>low)+n)/8 bytes) | lower bits (n*low bits) |
// payload (n*payloadBits bits). All bit vectors are LSB-first. Chunks are
// small (<=4096 entries) and always decoded whole, so there is no select
// structure: a chunk is one SST value read once.
package ef

import (
	"encoding/binary"
	"errors"
	"math/bits"
)

// MaxEntries is the chunk size the store cuts at: about one lookup block.
const MaxEntries = 4096

// Encode packs txnums (sorted ascending, may repeat) relative to base with
// one payloadBits-wide payload per entry. payloadBits 0 means no payload.
func Encode(base uint64, txnums []uint64, payload []byte, payloadBits int) []byte {
	n := len(txnums)
	var span uint64
	if n > 0 {
		span = txnums[n-1] - base
	}
	low := 0
	if n > 0 && span/uint64(n) > 0 {
		low = bits.Len64(span / uint64(n))
	}
	out := binary.AppendUvarint(nil, uint64(n))
	out = binary.AppendUvarint(out, uint64(low))
	out = binary.AppendUvarint(out, uint64(payloadBits))
	upperLen := int(span>>low) + n
	bw := writer{buf: out, bits: 0}
	for _, t := range txnums {
		hi := (t - base) >> low
		bw.set(int(hi) + bw.ones) // one bit per entry at position hi+i
		bw.ones++
	}
	bw.pad(upperLen)
	for _, t := range txnums {
		bw.put((t-base)&(1<<low-1), low)
	}
	for i := 0; i < n && payloadBits > 0; i++ {
		bw.put(uint64(payload[i]), payloadBits)
	}
	return bw.done()
}

var errCorrupt = errors.New("ef: corrupt chunk")

// Decode is the inverse of Encode. txnums are absolute (base added back).
func Decode(base uint64, v []byte) (txnums []uint64, payload []byte, err error) {
	pos := 0
	next := func() (uint64, error) {
		x, k := binary.Uvarint(v[pos:])
		if k <= 0 {
			return 0, errCorrupt
		}
		pos += k
		return x, nil
	}
	n64, err := next()
	if err != nil {
		return nil, nil, err
	}
	low64, err := next()
	if err != nil {
		return nil, nil, err
	}
	pb64, err := next()
	if err != nil {
		return nil, nil, err
	}
	n, low, pb := int(n64), int(low64), int(pb64)
	if n > MaxEntries || low > 64 || pb > 8 {
		return nil, nil, errCorrupt
	}
	r := reader{buf: v, bit: pos * 8}
	txnums = make([]uint64, n)
	var hi uint64
	for i := 0; i < n; i++ {
		for {
			b, ok := r.next()
			if !ok {
				return nil, nil, errCorrupt
			}
			if b {
				break
			}
			hi++
		}
		txnums[i] = hi
	}
	r.bit = (r.bit + 7) &^ 7 // upper vector is byte padded
	for i := range txnums {
		lo, ok := r.get(low)
		if !ok {
			return nil, nil, errCorrupt
		}
		txnums[i] = base + txnums[i]<<low + lo
	}
	if pb > 0 {
		payload = make([]byte, n)
		for i := range payload {
			p, ok := r.get(pb)
			if !ok {
				return nil, nil, errCorrupt
			}
			payload[i] = byte(p)
		}
	}
	return txnums, payload, nil
}

type writer struct {
	buf  []byte
	bits int // bits written past buf's uvarint header
	base int
	ones int
}

func (w *writer) set(i int) {
	if w.base == 0 {
		w.base = len(w.buf)
	}
	for len(w.buf) <= w.base+i/8 {
		w.buf = append(w.buf, 0)
	}
	w.buf[w.base+i/8] |= 1 << (i % 8)
	if i+1 > w.bits {
		w.bits = i + 1
	}
}

// pad closes the byte-aligned upper vector at exactly n bits.
func (w *writer) pad(n int) {
	if w.base == 0 {
		w.base = len(w.buf)
	}
	for len(w.buf) < w.base+(n+7)/8 {
		w.buf = append(w.buf, 0)
	}
	w.bits = len(w.buf) * 8
}

func (w *writer) put(x uint64, nbits int) {
	for k := 0; k < nbits; k++ {
		if w.bits%8 == 0 {
			w.buf = append(w.buf, 0)
		}
		if x>>k&1 == 1 {
			w.buf[w.bits/8] |= 1 << (w.bits % 8)
		}
		w.bits++
	}
}

func (w *writer) done() []byte { return w.buf }

type reader struct {
	buf []byte
	bit int
}

func (r *reader) next() (bool, bool) {
	if r.bit/8 >= len(r.buf) {
		return false, false
	}
	b := r.buf[r.bit/8]>>(r.bit%8)&1 == 1
	r.bit++
	return b, true
}

func (r *reader) get(nbits int) (uint64, bool) {
	var x uint64
	for k := 0; k < nbits; k++ {
		b, ok := r.next()
		if !ok {
			return 0, false
		}
		if b {
			x |= 1 << k
		}
	}
	return x, true
}

File: rs/store/src/ef.rs
//! store/ef: one posting chunk, plain Elias-Fano over txnums plus a fixed
//! width payload per entry. uvarint n | uvarint low | uvarint payloadBits |
//! upper bits (unary, byte padded) | low bits | payloads; all LSB-first.

use anyhow::{bail, Result};

pub const MAX_ENTRIES: usize = 4096;

struct BitWriter {
    buf: Vec<u8>,
    bits: usize,
    base: usize,
}
impl BitWriter {
    fn set(&mut self, i: usize) {
        while self.buf.len() <= self.base + i / 8 {
            self.buf.push(0);
        }
        self.buf[self.base + i / 8] |= 1 << (i % 8);
        if i + 1 > self.bits {
            self.bits = i + 1;
        }
    }
    fn pad(&mut self, n: usize) {
        while self.buf.len() < self.base + (n + 7) / 8 {
            self.buf.push(0);
        }
        self.bits = self.buf.len() * 8;
    }
    fn put(&mut self, x: u64, nbits: u32) {
        for k in 0..nbits {
            if self.bits % 8 == 0 {
                self.buf.push(0);
            }
            if x >> k & 1 == 1 {
                self.buf[self.bits / 8] |= 1 << (self.bits % 8);
            }
            self.bits += 1;
        }
    }
}

pub fn encode(base: u64, txnums: &[u64], payload: &[u8], payload_bits: u32) -> Vec<u8> {
    let n = txnums.len();
    let span = if n > 0 { txnums[n - 1] - base } else { 0 };
    let mut low = 0u32;
    if n > 0 && span / n as u64 > 0 {
        low = 64 - (span / n as u64).leading_zeros();
    }
    let mut out = Vec::new();
    crate::sst::put_uvarint(&mut out, n as u64);
    crate::sst::put_uvarint(&mut out, low as u64);
    crate::sst::put_uvarint(&mut out, payload_bits as u64);
    let upper_len = (span >> low) as usize + n;
    let base_off = out.len();
    let mut bw = BitWriter { buf: out, bits: 0, base: base_off };
    for (i, t) in txnums.iter().enumerate() {
        let hi = (t - base) >> low;
        bw.set(hi as usize + i);
    }
    bw.pad(upper_len);
    for t in txnums {
        bw.put((t - base) & ((1u64 << low) - 1), low);
    }
    if payload_bits > 0 {
        for p in payload.iter().take(n) {
            bw.put(*p as u64, payload_bits);
        }
    }
    bw.buf
}

struct BitReader<'a> {
    buf: &'a [u8],
    bit: usize,
}
impl BitReader<'_> {
    fn next(&mut self) -> Option<bool> {
        if self.bit / 8 >= self.buf.len() {
            return None;
        }
        let b = self.buf[self.bit / 8] >> (self.bit % 8) & 1 == 1;
        self.bit += 1;
        Some(b)
    }
    fn get(&mut self, nbits: u32) -> Option<u64> {
        let mut x = 0u64;
        for k in 0..nbits {
            if self.next()? {
                x |= 1 << k;
            }
        }
        Some(x)
    }
}

pub fn decode(base: u64, v: &[u8]) -> Result<(Vec<u64>, Vec<u8>)> {
    let mut pos = 0;
    let mut next = || -> Result<u64> {
        let (x, k) = crate::sst::uvarint(&v[pos..]).ok_or_else(|| anyhow::anyhow!("ef: corrupt chunk"))?;
        pos += k;
        Ok(x)
    };
    let n = next()? as usize;
    let low = next()? as u32;
    let pb = next()? as u32;
    if n > MAX_ENTRIES || low > 64 || pb > 8 {
        bail!("ef: corrupt chunk");
    }
    let mut r = BitReader { buf: v, bit: pos * 8 };
    let mut nums = vec![0u64; n];
    let mut hi = 0u64;
    for x in nums.iter_mut() {
        loop {
            match r.next() {
                None => bail!("ef: corrupt chunk"),
                Some(true) => break,
                Some(false) => hi += 1,
            }
        }
        *x = hi;
    }
    r.bit = (r.bit + 7) & !7;
    for x in nums.iter_mut() {
        let lo = r.get(low).ok_or_else(|| anyhow::anyhow!("ef: corrupt chunk"))?;
        *x = base + (*x << low) + lo;
    }
    let mut payload = Vec::new();
    if pb > 0 {
        for _ in 0..n {
            payload.push(r.get(pb).ok_or_else(|| anyhow::anyhow!("ef: corrupt chunk"))? as u8);
        }
    }
    Ok((nums, payload))
}
